src/Features/Core/Portable/MetadataAsSource/DecompilationMetadataAsSourceFileProvider.cs

```diff
@@ -103,112 +92,126 @@ internal sealed class DecompilationMetadataAsSourceFileProvider(IImplementationA
-                    skipWritingFile = true;
-                    break;
+                    var sourceFromMetadataService = temporaryDocument.Project.Services.GetRequiredService<IMetadataAsSourceService>();
+                    temporaryDocument = await sourceFromMetadataService.AddSourceToAsync(temporaryDocument, compilation, symbol, formattingOptions: null, cancellationToken).ConfigureAwait(false);
                 }
 
-                if (firstAttempt)
-                {
-                    firstAttempt = false;
-                }
-                else
+                // We have the content, so write it out to disk
+                var text = await temporaryDocument.GetValueTextAsync(cancellationToken).ConfigureAwait(false);
+
+                // Create the directory. It's possible a parallel deletion is happening in another process, so we may have
+                // to retry this a few times.
+                //
+                // If we still can't create the folder after 5 seconds, assume we will not be able to create it and
+                // continue without actually writing the text to disk.
+                var directoryToCreate = Path.GetDirectoryName(fileInfo.TemporaryFilePath)!;
+                var stopwatch = SharedStopwatch.StartNew();
+                var timeout = TimeSpan.FromSeconds(5);
+                var firstAttempt = true;
+                var skipWritingFile = false;
+
+                while (!IOUtilities.PerformIO(() => Directory.Exists(directoryToCreate)))
                 {
-                    await Task.Delay(DelayTimeSpan.Short, cancellationToken).ConfigureAwait(false);
-                }
+                    if (stopwatch.Elapsed > timeout)
+                    {
+                        // If we still can't create the folder after 5 seconds, assume we will not be able to create it.
+                        skipWritingFile = true;
+                        break;
+                    }
 
-                IOUtilities.PerformIO(() => Directory.CreateDirectory(directoryToCreate));
-            }
+                    if (firstAttempt)
+                    {
+                        firstAttempt = false;
+                    }
+                    else
+                    {
+                        await Task.Delay(DelayTimeSpan.Short, cancellationToken).ConfigureAwait(false);
+                    }
 
-            if (!skipWritingFile)
-            {
-                using (var textWriter = new StreamWriter(fileInfo.TemporaryFilePath, append: false, encoding: MetadataAsSourceGeneratedFileInfo.Encoding))
-                {
-                    text.Write(textWriter, cancellationToken);
+                    IOUtilities.PerformIO(() => Directory.CreateDirectory(directoryToCreate));
                 }
 
-                // Mark read-only
-                new FileInfo(fileInfo.TemporaryFilePath).IsReadOnly = true;
+                if (!skipWritingFile && !File.Exists(fileInfo.TemporaryFilePath))
```

Comment:
We're still under the !File.Exists() check you added on line 100 in the new version; isn't that part always true?

Reply:
I think this is guarding against the file going away in between (though there is technically a mutex held there so not sure if it could)?  I changed this in the followup though